File: config/src/test/scala/net/kemitix/thorp/config/ConfigQueryTest.scala
package net.kemitix.thorp.config

import java.nio.file.Paths
import java.util

import scala.jdk.CollectionConverters._
import net.kemitix.thorp.domain.Sources
import org.scalatest.FreeSpec

class ConfigQueryTest extends FreeSpec {

  def list(): util.List[ConfigOption]                     = List.empty.asJava
  def list(option: ConfigOption): util.List[ConfigOption] = List(option).asJava
  def list(option1: ConfigOption,
           option2: ConfigOption): util.List[ConfigOption] =
    List(option1, option2).asJava

  "show version" - {
    "when is set" - {
      "should be true" in {
        val result =
          ConfigQuery.showVersion(
            ConfigOptions.create(list(ConfigOption.version())))
        assertResult(true)(result)
      }
    }
    "when not set" - {
      "should be false" in {
        val result = ConfigQuery.showVersion(ConfigOptions.create(list()))
        assertResult(false)(result)
      }
    }
  }
  "batch mode" - {
    "when is set" - {
      "should be true" in {
        val result =
          ConfigQuery.batchMode(
            ConfigOptions.create(list(ConfigOption.batchMode())))
        assertResult(true)(result)
      }
    }
    "when not set" - {
      "should be false" in {
        val result = ConfigQuery.batchMode(ConfigOptions.empty())
        assertResult(false)(result)
      }
    }
  }
  "ignore user options" - {
    "when is set" - {
      "should be true" in {
        val result = ConfigQuery.ignoreUserOptions(
          ConfigOptions.create(list(ConfigOption.ignoreUserOptions())))
        assertResult(true)(result)
      }
    }
    "when not set" - {
      "should be false" in {
        val result = ConfigQuery.ignoreUserOptions(ConfigOptions.empty())
        assertResult(false)(result)
      }
    }
  }
  "ignore global options" - {
    "when is set" - {
      "should be true" in {
        val result = ConfigQuery.ignoreGlobalOptions(
          ConfigOptions.create(list(ConfigOption.ignoreGlobalOptions())))
        assertResult(true)(result)
      }
    }
    "when not set" - {
      "should be false" in {
        val result = ConfigQuery.ignoreGlobalOptions(ConfigOptions.empty())
        assertResult(false)(result)
      }
    }
  }
  "sources" - {
    val pathA = Paths.get("a-path")
    val pathB = Paths.get("b-path")
    "when not set" - {
      "should have current dir" in {
        val pwd      = Paths.get(System.getenv("PWD"))
        val expected = Sources.create(List(pwd).asJava)
        val result   = ConfigQuery.sources(ConfigOptions.empty())
        assertResult(expected)(result)
      }
    }
    "when is set once" - {
      "should have one source" in {
        val expected = Sources.create(List(pathA).asJava)
        val result =
          ConfigQuery.sources(
            ConfigOptions.create(list(ConfigOption.source(pathA))))
        assertResult(expected)(result)
      }
    }
    "when is set twice" - {
      "should have two sources" in {
        val expected = Sources.create(List(pathA, pathB).asJava)
        val result = ConfigQuery.sources(
          ConfigOptions.create(
            list(ConfigOption.source(pathA), ConfigOption.source(pathB))))
        assertResult(expected)(result)
      }
    }
  }

}

File: config/src/test/java/net/kemitix/thorp/config/ConfigQueryTest.java
package net.kemitix.thorp.config;

import net.kemitix.thorp.domain.Sources;
import org.assertj.core.api.WithAssertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class ConfigQueryTest
        implements WithAssertions {
    @Nested
    @DisplayName("show version")
    public class ShowVersionTest{
        @Test
        @DisplayName("when set then show")
        public void whenSet_thenShow() {
            assertThat(ConfigQuery.showVersion(
                    ConfigOptions.create(
                            Collections.singletonList(
                                    ConfigOption.version()))
            )).isTrue();
        }
        @Test
        @DisplayName("when not set then do not show")
        public void whenNotSet_thenDoNotShow() {
            assertThat(ConfigQuery.showVersion(
                    ConfigOptions.create(
                            Collections.emptyList())
            )).isFalse();
        }
    }
    @Nested
    @DisplayName("batch mode")
    public class BatchModeTest{
        @Test
        @DisplayName("when set then show")
        public void whenSet_thenShow() {
            assertThat(ConfigQuery.batchMode(
                    ConfigOptions.create(
                            Collections.singletonList(
                                    ConfigOption.batchMode()))
            )).isTrue();
        }
        @Test
        @DisplayName("when not set then do not show")
        public void whenNotSet_thenDoNotShow() {
            assertThat(ConfigQuery.batchMode(
                    ConfigOptions.create(
                            Collections.emptyList())
            )).isFalse();
        }
    }
    @Nested
    @DisplayName("ignore user options")
    public class IgnoreUserOptionsTest{
        @Test
        @DisplayName("when set then show")
        public void whenSet_thenShow() {
            assertThat(ConfigQuery.ignoreUserOptions(
                    ConfigOptions.create(
                            Collections.singletonList(
                                    ConfigOption.ignoreUserOptions()))
            )).isTrue();
        }
        @Test
        @DisplayName("when not set then do not show")
        public void whenNotSet_thenDoNotShow() {
            assertThat(ConfigQuery.ignoreUserOptions(
                    ConfigOptions.create(
                            Collections.emptyList())
            )).isFalse();
        }
    }
    @Nested
    @DisplayName("ignore global options")
    public class IgnoreGlobalOptionsTest{
        @Test
        @DisplayName("when set then show")
        public void whenSet_thenShow() {
            assertThat(ConfigQuery.ignoreGlobalOptions(
                    ConfigOptions.create(
                            Collections.singletonList(
                                    ConfigOption.ignoreGlobalOptions()))
            )).isTrue();
        }
        @Test
        @DisplayName("when not set then do not show")
        public void whenNotSet_thenDoNotShow() {
            assertThat(ConfigQuery.ignoreGlobalOptions(
                    ConfigOptions.create(
                            Collections.emptyList())
            )).isFalse();
        }
    }
    @Nested
    @DisplayName("source")
    public class SourcesTest {
        Path pathA = Paths.get("a-path");
        Path pathB = Paths.get("b-path");
        @Test
        @DisplayName("when not set then use current directory")
        public void whenNoSet_thenCurrentDir() {
            Sources expected = Sources.create(
                    Collections.singletonList(
                            Paths.get(
                                    System.getenv("PWD")
                            )));
            assertThat(ConfigQuery.sources(ConfigOptions.empty()))
                    .isEqualTo(expected);
        }
        @Test
        @DisplayName("when one source then have one source")
        public void whenOneSource_thenOneSource() {
            List<Path> expected = Collections.singletonList(pathA);
            assertThat(ConfigQuery.sources(
                    ConfigOptions.create(
                            Collections.singletonList(
                                    ConfigOption.source(pathA)))).paths())
                    .isEqualTo(expected);
        }
        @Test
        @DisplayName("when two sources then have two sources")
        public void whenTwoSources_thenTwoSources() {
            List<Path> expected = Arrays.asList(pathA, pathB);
            assertThat(
                    ConfigQuery.sources(
                            ConfigOptions.create(
                                    Arrays.asList(
                                            ConfigOption.source(pathA),
                                            ConfigOption.source(pathB))
                            )).paths())
                    .isEqualTo(expected);
        }
    }
}
